### ISSUE TITLE
Leaks passwords/secrets to the operating system

### ISSUE THREAD

**Issue by michaelfranzl** [cmer/nobspw#6]
Arguments to processes (here: `grep`) are world-readable (at least in the Linux kernel) and can be inspected, for example, via `ps -ejHf`. Thus, the following line leaks the passwords/secrets to all users (or even loggers!) of the OS:

https://github.com/cmer/nobspw/blob/efd01ada94484e982d16023568d9c65154a45368/lib/nobspw/password_checker.rb#L41

**Comment by cmer** [comment]
Good point. Thanks. What do you think is a valid fix?

**Comment by johannesluedke** [comment]
One could put the passwords in a database, and then has to make sure that the database query is excluded from logs. This would make the setup for the user of this gem more complicated though

**Comment by johannesluedke** [comment]
Why not do something along the lines of this?

```
File.foreach(path) do |file|
  file.grep(/^#{@password}$/)
end
```

**Comment by johannesluedke** [comment]
@cmer I added a PR for that https://github.com/cmer/nobspw/pull/10. The issue with this approach might be that it's too slow

**Comment by cmer** [comment]
Awesome @johannesluedke! Have you benchmarked the two methods by any chance? I worry that Ruby grep might be incredibly slow. Would be great to know at least what the tradeoffs are (if any).

Thanks again!

**Comment by cmer** [comment]
I just ran some benchmarks to see the impact. The Ruby version is **significantly** slower. I wonder if there's an alternative that doesn't leak and uses gnu grep...

```
       user     system      total        real
Ruby  0.773207   0.037857   0.811064 (  0.815542)
Shell  0.000026   0.000005   0.000031 (  0.000029)
```

[https://github.com/cmer/nobspw/blob/master/misc/grep_benchmark.rb](https://github.com/cmer/nobspw/blob/master/misc/grep_benchmark.rb)

**Comment by johannesluedke** [comment]
@cmer But these numbers are in seconds, for 100 iterations, so we're talking about ~77ms, right? Maybe one could still make this the default, and offer a configuration option to turn common_phrase check off? 

One could offer these other option, but it requires some setup and only works if there is a database in the project (which is probably the case, nobspw will likely be used when there are users that are stored in the database): import dictionary with INSERT INTO and monkey-patch `too_common_phrase?` 

```ruby
module NOBSPW
  module ValidationMethods
    private

    def password_too_common?
      # Security Note: Be aware that when this query would be logged, it would
      # log the password in cleartext.
      Rails.logger.silence do
         ::PasswordDict.exists?(@password)
      end
    end
  end
end
```

**Comment by cmer** [comment]
I am an imbecile. My shell-out benchmarks were fast because I was actually not even shelling out. 😂 I was using quotes instead of backticks.

Here are the real benchmarks:

```
                        user     system      total        real
Shell               0.014936   1.187746   4.839292 (  4.982993)
Ruby Grep           2.781987   0.060514   2.842501 (  2.849254)
Ruby Loop           2.979807   0.037881   3.017688 (  3.025581)
Open3 stdin         0.032080   1.398143   4.916618 (  4.994484)
Subprocess stdin    0.044234   1.383056   4.842164 (  4.922027)
```

It does look like Ruby Grep is actually pretty fast. I'll update the code shortly. Thanks again!

**Comment by cmer** [comment]
Already. I have released v0.6.0.

It no longer leaks passwords to `ps`, while still shelling out to `grep`, simply because it's faster than using Ruby grep. The password is now piped to `grep` via `stdin` instead of being passed as an argument like before.

There is now a configuration setting called `use_ruby_grep` if you prefer to use Ruby grep. `use_ruby_grep` is `false` by default. If the `grep` command is not found (ie: not in path), it gracefully falls back to Ruby grep.

This should offer optimal flexibility, security and speed.
